### PR TITLE
Fix 3DS Authorize and Capture processing with Deferred Intent

### LINKED ISSUE #2764: [Deferred intent] Card - Handle capture-later transactions
### Acceptance criteria

- The following critical flow [Merchant -> Manual capture](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows-using-the-classic-checkout-experience#manual-capture) must work as described
- It must work with both regular and 3DS cards
- Errors must be handled gracefully:
  - A relevant, user-friendly message must be shown to shoppers
  - A relevant note must be added to the order
  - A detailed message of the error must be logged to the log files if enabled
  - The order status must reflect the payment status

### PR DESCRIPTION
**Pull request by james-allan**
Fixes #2764

## Changes proposed in this Pull Request:

This PR fixes an issue using 3DS cards with Auth and Capture enabled using our split UPE with deferred intent implementation so far on the `add/deferred-intent` branch. 

If you have auth and capture enabled, and process a payment using a SCA/3DS enabled card, the orders notes on the order will be missing the charge ID, and the order will be set to on-hold before the 3DS verification has been completed. This PR fixes these two issues.

## Testing instructions

1. Enable UPE and auth/capture in the Stripe settings. 
   - **WooCommerce → Settings → Payments → Stripe → Settings**
   - Enable the _'Issue an authorization on checkout, and capture later'_ setting
2. Place any simple product in your cart and enter a 3DS payment card number. eg `4000002500003155`
3. Don't complete the auth. 
4. Go to admin and view the order. On `add/deferred-intent`, you'll notice two things:
   1. The order note has no charge ID
   2. The order has an on-hold status before you've passed the 3DS auth.

| Order status | Order notes |
|--------|--------|
| Screenshot 2023-12-06 at 5 00 54 pm | Screenshot 2023-12-06 at 4 57 44 pm | 

5. You can complete the auth, the rest of the process for auth and capture works as expected.
6. Repeat the steps above on this branch you'll notice that
   - Unlike on `add/deferred-intent`, while the checkout hangs with the auth modal being shown, the order won't update it's status. It will remain pending payment.
   - There will be an order note about the payment intent being created but no reference to a charge yet.
7. Complete authorization and review the order again. 
8. It will now have an on-hold status and the order note with a charge ID. 
9. Verify this change doesn't break any other flow. 

---

-   [ ] Covered with tests (or have a good reason not to test in description ☝️)
-   [ ] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [ ] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)

### CODE REVIEW

File: includes/payment-methods/class-wc-stripe-payment-request.php [hunk 1]
```diff
@@ -1211,6 +1211,7 @@ public function get_shipping_options( $shipping_address, $itemized_display_items
 			if ( ! empty( $packages ) && WC()->customer->has_calculated_shipping() ) {
 				foreach ( $packages as $package_key => $package ) {
 					if ( empty( $package['rates'] ) ) {
+						print_r( $package, true );
```
Comment (mattallan) [inline includes/payment-methods/class-wc-stripe-payment-request.php]: Looks like some debugging code made it into this PR 😅

Comment (james-allan) [reply includes/payment-methods/class-wc-stripe-payment-request.php]: Yeah I was attempting debug the failing test, I had put this PR on-hold until I had worked out what to do. Turns out it's a WC core issue: ref p1701989810708879/1701234236.117869-slack-C055WHLA98D

File: client/blocks/upe/payment-method.js
Comment (james-allan) [inline client/blocks/upe/payment-method.js:1]: I've deleted this file given it is no longer being used.

**Review by mattallan — APPROVED** [review]
Nice changes @james-allan !


> Intents with no charge likely require further action like 3DS

When I read this inline comment, I was wondering if we could be more specific about this rather than assuming payment intents without a charge are likely 3DS, i.e. could we check the status of the `$payment_intent` here?

The more I thought about it the more I didn't think it was necessary because we're already checking the status of the intent later on. I've approved the PR but curious to get your thoughts on this :)

**Comment by james-allan** [comment]
> When I read this inline comment, I was wondering if we could be more specific about this rather than assuming payment intents without a charge are likely 3DS, i.e. could we check the status of the $payment_intent here?

Yeah so this logic at its base level is saying we can only process the charge response if it has a charge. The comment I included about it _likely_ being a 3DS scenario was to add at least 1 reason why it might not include a charge.

I had a look at how WooPayments handles this and they have a very different way of handling intent responses. They have two functions which handle specific parts: 

- `attach_intent_info_to_order()`
- `update_order_status_from_intent()`

The Stripe plugin only has a single `process_response()` which does both and expects the first parameter to be a charge object. If the charge doesn't exist, we were previously passing that function a `bool` and that was leading to errors like this:

```
PHP Warning:  Attempt to read property "id" on bool in /wp-content/plugins/woocommerce-gateway-stripe/includes/abstracts/abstract-wc-stripe-payment-gateway.php on line 528
```

So, I think this logic is correct in that it checks if there's a charge object before calling a function that expects a charge object.

**Pull request merged**